### ISSUE TITLE
Apparent maximum number of nodes in a graph

### ISSUE THREAD

**Issue by a-musing-moose** [python-bonobo/bonobo#271]
Hi Devs,

I seemed to be having an odd issue. I have a `Graph` which has grown to a fair number of nodes (13 at this time). Running it locally on my machine it works fine. As soon as I run it on production I get some strange behaviour:

I would expect all nodes to be initialised and start processing data.  

What actually happened is that only 10 nodes started. The remaining did not get any input (and were missing the `+` at the beginning).  Those nodes which were successfully started seemed to process less and less data as they went through the graph until it just stopped doing anything. Frozen state output looks like this:

```
 - set_custom_where in=1 out=1 [done] 
 + extract_from_mysql in=1 out=81930  
 + combine_group in=73737 out=73737  
 + attach_labels in=65544 out=65544  
 + set_hash_id in=57351 out=57351  
 + add_alias in=49158 out=49158  
 + process_created_date in=40965 out=40965  
 + process_last_modified_date in=32772 out=32772  
 + process_birthday in=24579 out=24579  
 + process_country_id_to_code in=16386 out=16386  
 + add_first_address_phone_and_code_to_person in=8193 out=8193  
   inject_person_uuid   
   push 
```

What I *think* is happening is that the default executor is `concurrent.futures.ThreadpoolWorker`. The first parameter of which is `max_workers` which `bonobo` currently leaves as the default, `None`. The Python docs say that if it is left as `None` then `max_workers = num_cpus * 5`. 

My local machine has 4 cores, thus 20 workers. The production machine only has 2, therefore is caps the number of workers at 10.

The problem resolves itself if there are a small number of records going through, as the earlier nodes close and the later ones get started. But I think there is a max buffer size between nodes, and as such if a lot of data is passing through, the earlier steps never complete and the whole pipeline locks up.
 
As a work around I am currently sub-classing the executor and adding a large value for max_workers, then passing that custom executor to `run`.

## Versions

* Bonobo version: `0.6.2`

* Python version:
```
CPython 3.6.4 (default, Jan 27 2018, 09:58:56) 
[GCC 7.2.0]
```

* Platform:

`Linux pretty-on-the-inside 4.15.0-23-generic #25-Ubuntu SMP Wed May 23 18:02:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
   
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04 LTS"
```

**Comment by a-musing-moose** [comment]
This also effects the `ProcessPoolExecutorStrategy`.

A naive approach would be to pass the `Graph` to `create_executor` and simply count the number of nodes to determine `max_workers`.

**Comment by hartym** [comment]
Thanks for the report. It's indeed a very bad bug. The problem is a deadlock, happening when too much queues are full. Not sure we need more workers, will think about it within a few days.

**Comment by hartym** [comment]
Can you try https://github.com/python-bonobo/bonobo/commit/3e2444ccb69d6c9aa394a0de0a987260c8bd2a3d and confirm whether or not it fixes the problem? If it does fix it, I'll release a new 0.6 version with this. Thanks!

**Comment by hartym** [comment]
Should be fixed in 0.6.3, please reopen if it's not the case for you.
